transport/src/main/java/io/netty/channel/DefaultChannelHandlerContext.java

```diff
@@ -423,248 +422,183 @@ void invokeChannelWritabilityChanged() {
     }
 
     @Override
     public Future<Void> deregister() {
-        return deregister(newPromise());
-    }
-
-    @Override
-    public Future<Void> bind(final SocketAddress localAddress, final Promise<Void> promise) {
-        requireNonNull(localAddress, "localAddress");
-        if (isNotValidPromise(promise)) {
-            // cancelled
-            return promise;
-        }
-
         EventExecutor executor = executor();
         if (executor.inEventLoop()) {
-            findAndInvokeBind(localAddress, promise);
-        } else {
-            safeExecute(executor, () -> findAndInvokeBind(localAddress, promise), promise, null);
+            return findAndInvokeDeregister();
         }
+        Promise<Void> promise  = newPromise();
+        safeExecute(executor, () -> PromiseNotifier.fuse(findAndInvokeDeregister(), promise), promise, null);
         return promise;
     }
-
-    private void findAndInvokeBind(SocketAddress localAddress, Promise<Void> promise) {
+    private Future<Void> findAndInvokeBind(SocketAddress localAddress) {
         DefaultChannelHandlerContext ctx = findContextOutbound(MASK_BIND);
         if (ctx == null) {
-            failRemoved(this, promise);
-            return;
+            return failRemoved(this);
         }
-        ctx.invokeBind(localAddress, promise);
+        return ctx.invokeBind(localAddress);
     }
 
-    private void invokeBind(SocketAddress localAddress, Promise<Void> promise) {
+    private Future<Void> invokeBind(SocketAddress localAddress) {
         try {
-            handler().bind(this, localAddress, promise);
+            return handler().bind(this, localAddress);
         } catch (Throwable t) {
-            handleOutboundHandlerException(t, false, promise);
+            return handleOutboundHandlerException(t, false);
         }
     }
 
-    @Override
-    public Future<Void> connect(SocketAddress remoteAddress, Promise<Void> promise) {
-        return connect(remoteAddress, null, promise);
-    }
-
     @Override
     public Future<Void> connect(
-            final SocketAddress remoteAddress, final SocketAddress localAddress, final Promise<Void> promise) {
+            final SocketAddress remoteAddress, final SocketAddress localAddress) {
         requireNonNull(remoteAddress, "remoteAddress");
-        if (isNotValidPromise(promise)) {
-            // cancelled
-            return promise;
-        }
-
         EventExecutor executor = executor();
         if (executor.inEventLoop()) {
-            findAndInvokeConnect(remoteAddress, localAddress, promise);
-        } else {
-            safeExecute(executor, () -> findAndInvokeConnect(remoteAddress, localAddress, promise), promise, null);
+            return findAndInvokeConnect(remoteAddress, localAddress);
         }
+        Promise<Void> promise  = newPromise();
+            safeExecute(executor, () ->
+                    PromiseNotifier.fuse(findAndInvokeConnect(remoteAddress, localAddress), promise), promise, null);
```

Comment:
Indentation is off.

Reply:
hmm looks correct to me.

Reply:
I mean, the call to `safeExecute(…` should be aligned with the `Promise<Void> promise = …` declaration on the line before.

Reply:
ah gotcha